torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,212 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+#ifdef ncclGroupStart
+#define NCCL_VERSION 2
+#endif
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, std::unordered_map<int, ncclComm_t> > _communicators;
+std::vector<ncclComm_t *> _all_communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto comm: _all_communicators) {
+    ncclCommDestroy(*comm);
+  }
+  Py_RETURN_NONE;
+}
+
+static std::unordered_map<int, ncclComm_t> _get_communicator(const int *devs, int ndevices) {
```

Comment:
1. Return a reference from this function. Right now it's doing a ton of copies.
2. It would by much simpler if you used `IntList` from ATen to represent a list of devices and added a custom hash function to `_communicators`.